File: vripper-server/src/main/java/tn/mnlr/vripper/services/GlobalStateService.java
package tn.mnlr.vripper.services;

import lombok.Getter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.EnableScheduling;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import tn.mnlr.vripper.download.DownloadService;
import tn.mnlr.vripper.event.Event;
import tn.mnlr.vripper.event.EventBus;
import tn.mnlr.vripper.services.domain.GlobalState;

@Service
@EnableScheduling
public class GlobalStateService {

  private final DownloadService downloadService;
  private final DataService dataService;
  private final EventBus eventBus;
  @Getter private GlobalState currentState;

  @Autowired
  public GlobalStateService(
      DownloadService downloadService, DataService dataService, EventBus eventBus) {
    this.downloadService = downloadService;
    this.dataService = dataService;
    this.eventBus = eventBus;
  }

  @Scheduled(fixedDelay = 3000)
  private void interval() {
    GlobalState newGlobalState =
        new GlobalState(
            downloadService.runningCount(),
            downloadService.pendingCount(),
            dataService.countErrorImages());
    if (!newGlobalState.equals(currentState)) {
      currentState = newGlobalState;
      eventBus.publishEvent(Event.wrap(Event.Kind.GLOBAL_STATE, currentState));
    }
  }
}

File: vripper-core/src/main/kotlin/me/mnlr/vripper/services/GlobalStateService.kt
package me.mnlr.vripper.services

import kotlinx.coroutines.*
import me.mnlr.vripper.download.DownloadService
import me.mnlr.vripper.event.Event
import me.mnlr.vripper.event.EventBus
import me.mnlr.vripper.formatSI
import me.mnlr.vripper.model.GlobalState
import java.util.concurrent.atomic.AtomicLong

class GlobalStateService(
    private val downloadService: DownloadService,
    private val vgAuthService: VGAuthService,
    private val eventBus: EventBus,
    private val dataTransaction: DataTransaction
) {

    companion object {
        const val DOWNLOAD_POLL_RATE = 250
    }

    private var currentState: GlobalState = newValue()
    private val coroutineScope = CoroutineScope(Dispatchers.Default)
    private val bytesCount = AtomicLong(0)
    private var currentValue = 0L

    fun init() {
        coroutineScope.launch {
            while (isActive) {
                val newValue = bytesCount.getAndSet(0)
                currentValue = newValue
                delay(DOWNLOAD_POLL_RATE.toLong())
            }
        }

        coroutineScope.launch {
            while (isActive) {
                val newGlobalState = newValue()
                if (newGlobalState != currentState) {
                    currentState = newGlobalState
                    eventBus.publishEvent(Event(Event.Kind.DOWNLOAD_STATUS, currentState))
                }
                delay(200)
            }
        }
    }

    fun reportDownloadedBytes(count: Long) {
        this.bytesCount.addAndGet(count)
    }

    private fun newValue(): GlobalState {
        return GlobalState(
            downloadService.runningCount(),
            downloadService.pendingCount(),
            dataTransaction.countImagesInError(),
            vgAuthService.loggedUser,
            ((currentValue * 1000) / DOWNLOAD_POLL_RATE).formatSI()
        )
    }

    fun get(): GlobalState {
        return currentState.copy()
    }
}
